src/gui/symbology/qgsextentbufferdialog.cpp

```diff
@@ -0,0 +1,128 @@
+/***************************************************************************
+    qgsextentbufferdialog.cpp
+    ---------------------
+    begin                : December 2024
+    copyright            : (C) 2024 by Juho Ervasti
+    email                : juho dot ervasti at gispo dot fi
+ ***************************************************************************
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+
+#include "qgsextentbufferdialog.h"
+#include "moc_qgsextentbufferdialog.cpp"
+#include "qdialogbuttonbox.h"
+#include "qgsexpressioncontext.h"
+#include "qgshelp.h"
+#include "qgspanelwidget.h"
+#include "qgssymbol.h"
+#include "qgssymbolwidgetcontext.h"
+#include "qgsvectorlayer.h"
+
+QgsExtentBufferWidget::QgsExtentBufferWidget( QgsSymbol *symbol, QgsVectorLayer *layer, QWidget *parent )
+  : QgsPanelWidget( parent ), mSymbol( symbol ), mLayer( layer )
+{
+  setupUi( this );
+
+  mExtentBufferSpinBox->setValue( mSymbol-> extentBuffer() );
+
+  connect( mExtentBufferSpinBox, static_cast < void ( QgsDoubleSpinBox::* )( double ) > ( &QgsDoubleSpinBox::valueChanged ), this, [ = ]()
+  {
+    emit widgetChanged();
+  } );
+
+  registerDataDefinedButton( mExtentBufferDDButton, QgsSymbol::Property::ExtentBuffer );
+}
+
+QgsSymbolWidgetContext QgsExtentBufferWidget::context() const
+{
+  return mContext;
+}
+
+void QgsExtentBufferWidget::setContext( const QgsSymbolWidgetContext &context )
+{
+  mContext = context;
+}
+
+void QgsExtentBufferWidget::registerDataDefinedButton( QgsPropertyOverrideButton *button, QgsSymbol::Property key )
+{
+  // pass in nullptr to avoid id, feature and geometry variables being added
+  // since the buffer is not evaluated per-feature
+  button->init( static_cast< int >( key ), mSymbol->dataDefinedProperties(), QgsSymbol::propertyDefinitions(), nullptr );
+  connect( button, &QgsPropertyOverrideButton::changed, this, [ = ]()
+  {
+    emit widgetChanged();
+  } );
+
+  button->registerExpressionContextGenerator( this );
+}
+
+QgsExpressionContext QgsExtentBufferWidget::createExpressionContext() const
+{
+  QList<QgsExpressionContextScope *> scopes = mContext.globalProjectAtlasMapLayerScopes( mLayer );
```

Comment:
You'll also need to handle QgsSymbolWidgetContext if you want access to the canvas in order to get the @map_scale variable

Reply:
@nyalldawson 

Is it not sufficiently handled? When I'm testing I can use the @map_scale etc. variables. `mContext` here is a `QgsSymbolWidgetContext` object and it is set for the widget when it is created in `QgsSymbolListWidget`. Is there something else that needs to be done?

Reply:
Have a look at eg how QgsTextFormatWidget::createExpressionContext handles this -- the logic here should be almost identical